include/fuzzuf/executor/traits/afl_capable.hpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * fuzzuf
+ * Copyright (C) 2022 Ricerca Security
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+/**
+ * @file afl_capable.hpp
+ * @author Ricerca Security <fuzzuf-dev@ricsec.co.jp>
+ */
+
+#ifndef FUZZUF_INCLUDE_EXECUTOR_TRAITS_AFL_CAPABLE_HPP
+#define FUZZUF_INCLUDE_EXECUTOR_TRAITS_AFL_CAPABLE_HPP
+
+#include "fuzzuf/executor/native_linux_executor.hpp"
+#include "fuzzuf/feedback/exit_status_feedback.hpp"
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+/**
+ * @class DynAFLCapable
+ * @brief Represents minimal requirements for an AFL-capable executor.
+ * 
+ * @details The executor for AFL must have the methods declared in this class.
+ * This class is to perform type erasure for the executor class to abstract
+ * the executor on the algortihms. We found that the boost's type_erasure does
+ * not meet our needs, so we implemented our own type erasure.
+ */
+class DynAFLCapable
+{
+public:
+  template<class T> DynAFLCapable(const T& executor)
+  : _container(new DynContainerDerived<T>(executor))
+  {}
+  ~DynAFLCapable() { delete _container; }
+
+  DynAFLCapable(const DynAFLCapable&) = delete;
+  DynAFLCapable(DynAFLCapable&&) = delete;
+  DynAFLCapable &operator=(const DynAFLCapable&) = delete;
+  DynAFLCapable &operator=(DynAFLCapable&&) = delete;
+  DynAFLCapable() = delete;
+
+  void Run(const u8 *buf, u32 len, u32 timeout_ms=0) {
+    _container->Run(buf, len, timeout_ms);
+  }
+
+  InplaceMemoryFeedback GetAFLFeedback() {
+    return _container->GetAFLFeedback();
+  }
+
+  ExitStatusFeedback GetExitStatusFeedback() {
+    return _container->GetExitStatusFeedback();
+  }
+
+  void ReceiveStopSignal() {
+    _container->ReceiveStopSignal();
+  }
+
+private:
+  class DynContainerBase {
+  public:
+    virtual ~DynContainerBase() {}
+    virtual void Run(const u8 *buf, u32 len, u32 timeout_ms=0) = 0;
+    virtual InplaceMemoryFeedback GetAFLFeedback() = 0;
+    virtual ExitStatusFeedback GetExitStatusFeedback() = 0;
+    virtual void ReceiveStopSignal() = 0;
+  };
+
+  template<class T>
+  class DynContainerDerived : public DynContainerBase {
+  public:
+    DynContainerDerived(T const &executor) : _container(executor) {}
+
+    void Run(const u8 *buf, u32 len, u32 timeout_ms=0) {
+      _container->Run(buf, len, timeout_ms);
```

Comment:
Should the type of `DynContainerDerived<T>::_container` be pointer types?
If so, it must be noted in comments.
Also, if you can restrict the type (for example, allow only `std::shared_ptr<Foo>` as `T`), it would be better.
Accepting arbitrary pointer types as `T` means that we want to use at least two types among `Exector*`, `std::unique_ptr<Executor>`, and `std::shared_ptr<Executor>`.

Reply:
Fixed with [6532c86](https://github.com/fuzzuf/fuzzuf/pull/41/commits/6532c867cab29106390551e6f62aa905652fbbfd) and [16bb5ea](https://github.com/fuzzuf/fuzzuf/pull/41/commits/16bb5eaadfce9f423d019f2c9a6187b9ac375998) by restricting the type with `std::shared_ptr<T>`.